### ISSUE TITLE
Custom module - Using create-files checks [xml] configuration before deploying [files]

### ISSUE THREAD

**Issue by boughtonp** [eclipse/jetty.project#3531]
I am creating a custom module, but when attempting the first deploy it errors, because the files specified in the [xml] section do not yet exist.

Module Configuration:
```
[xml]
etc/mymodule-rewrite-rules.xml

[files]
basehome:modules/mymodule
```

Command run:
`java -jar ../jetty-distribution-9.4.15.v20190215/start.jar --create-files`

Result:
```
java.io.IOException: Cannot read file: etc/mymodule-rewrite-rules.xml
        at org.eclipse.jetty.start.StartArgs.addUniqueXmlFile(StartArgs.java:223)
        at org.eclipse.jetty.start.StartArgs.expandModules(StartArgs.java:541)
        at org.eclipse.jetty.start.Main.processCommandLine(Main.java:351)
        at org.eclipse.jetty.start.Main.main(Main.java:75)
```

Commenting out the xml block allows the files to be copied, but of course the rewrite rules are then not enabled.

This configuration exists within the built-in modules, but because their config files already exists in jetty-home this issue doesn't apply.

Is it valid to use the xml section for custom config files (in which case this is a bug), or is there another option for deploying/enabling?

**Comment by sbordet** [comment]
I'm not sure I understand.

If you have your custom module and you have your custom XML, then both must be present in `$JETTY_BASE`, respectively under `$JETTY_BASE/modules` and `$JETTY_BASE/etc`.

The `create-files` command applies to the `[files]` section, where you can add directories and files to `$JETTY_BASE/lib` - for example for dependencies that your custom module requires to work.

**Comment by boughtonp** [comment]
I'd like the module to be self-contained so it can just be dropped in the modules directory and it will configure `$JETTY_BASE` accordingly.

So, given an initial file structure looking like this:

```
jetty-base/
  start.ini
  modules/
    mymodule.mod
    mymodule/
      etc/
        mymodule-rewrite-rules.xml
```

After running `create-files` I would expect it to be:

```
jetty-base/
  start.ini
  etc/
    mymodule-rewrite-rules.xml
  modules/
    mymodule.mod
    mymodule/
      etc/
        mymodule-rewrite-rules.xml
```

And that works - but only with the `[xml]` section in the mymodule.mod commented out.

With the `[xml]` section enabled, it looks for that file before it has been copied, and thus errors.


Basically, it comes down to lines 535..549 of StartArgs.java

https://github.com/eclipse/jetty.project/blob/jetty-9.4.x/jetty-start/src/main/java/org/eclipse/jetty/start/StartArgs.java#L535

If the "Register Download operations" block is moved before "Find and Expand XML files" then it will work the way I expect - but I don't know whether the positioning is deliberate or just due to order of implementation.

**Comment by sbordet** [comment]
I don't think we hardcode XMLs to be under `$JETTY_BASE/etc`.

If your module has:

```
[xml]
mymodule/etc/mymodule-rewrite-rules.xml
```

does not work?

I don't see why you need to copy the XML file.

Also I don't understand what you mean by "self-contained": having a *.mod file then the `mymodule` directory, then an XML file under `mymodule` is not more self-contained than having a *.mod file and an XML file under the `etc` directory?

**Comment by boughtonp** [comment]
> I don't see why you need to copy the XML file.

I had assumed that was necessary, but you're right - I now have it worked without copying using:
```
[xml]
modules/mymodule/etc/mymodule-rewrite-rules.xml
```

**Comment by sbordet** [comment]
@boughtonp I still think that it would be better if you respect the convention of having XML files under `$JETTY_BASE/etc`.
Having said that, if you're happy with your current solution, feel free to close the issue :smiley:

**Comment by boughtonp** [comment]
At this stage it's convenient having it all in the module dir, but I do intend to follow with the grain of how Jetty works so I probably will go with `$JETTY_BASE/etc` once I have everything else done.
